File: TEMA 4/riego.pl
% -------------------------------
% SISTEMA DE RIEGO MEJORADO
% -------------------------------

% ========== BASE DE CONOCIMIENTO ==========
% Hechos dinámicos para permitir modificaciones en tiempo real
:- dynamic humedad_suelo/1, temperatura/1, hora/1, pronostico_lluvia/1.

% Datos iniciales (pueden cambiarse con assert/retract)
humedad_suelo(baja).
temperatura(35).
hora(20).
pronostico_lluvia(false).

% ========== REGLAS MEJORADAS ==========
% Hora adecuada para riego (con validación)
hora_adecuada :-
    hora(H),
    number(H),      % Verifica que sea un número
    between(0,23,H), % Hora válida (0-23)
    (H < 10 ; H > 18).

% Regla principal mejorada con validaciones
activar_riego :-
    humedad_suelo(baja),
    pronostico_lluvia(false),
    hora_adecuada,
    \+ alerta_calor. % No activar si hay alerta de calor

% Sistema de alertas mejorado
alerta_calor :-
    temperatura(T),
    number(T),
    T >= 32. % Temperatura peligrosa

alerta_sequia :-
    humedad_suelo(muy_baja),
    temperatura(T), T > 30.

% ========== SISTEMA DE RECOMENDACIONES ==========
recomendacion :-
    activar_riego,
    alerta_calor, !,
    writeln('ALERTA: Temperatura peligrosa detectada'),
    writeln('RECOMENDACION: Postergue riego o use sistema por goteo nocturno').

recomendacion :-
    alerta_sequia, !,
    writeln('ALERTA CRITICA: Sequia extrema detectada'),
    writeln('RECOMENDACION: Riego de emergencia inmediato').

recomendacion :-
    activar_riego, !,
    writeln('RIEGO AUTORIZADO: Condiciones óptimas detectadas').

recomendacion :-
    writeln('SISTEMA EN ESPERA: No se requieren acciones').

% ========== INTERFAZ DE DIAGNÓSTICO ==========
estado_riego('SISTEMA ACTIVADO') :- activar_riego.
estado_riego('SISTEMA EN ESPERA') :- \+ activar_riego.

% ========== HELPERS PARA PRUEBAS ==========
% Predicado para mostrar todos los sensores
estado_sensores :-
    writeln('\n=== ESTADO ACTUAL DE SENSORES ==='),
    (humedad_suelo(H) -> format('Humedad suelo: ~w~n', [H]) ; writeln('Humedad: No definida')),
    (temperatura(T) -> format('Temperatura: ~w°C~n', [T]) ; writeln('Temperatura: No definida')),
    (hora(H) -> format('Hora actual: ~w:00~n', [H]) ; writeln('Hora: No definida')),
    (pronostico_lluvia(L) -> 
        (L == true -> writeln('Pronostico: Lluvia esperada') ; writeln('Pronostico: No hay lluvia'))
     ; writeln('Pronostico: No definido')),
    writeln('==============================\n').

File: AA 4.2 REPORTE PRACTICO/riego.py
# sistema_riego.py
from typing import Literal

# Tipos posibles
Humedad = Literal["baja", "muy_baja", "media", "alta"]

def hora_adecuada(hora: int) -> bool:
    """Devuelve True si la hora es adecuada para regar (antes de las 10 o después de las 18)."""
    return 0 <= hora <= 23 and (hora < 10 or hora > 18)

def alerta_calor(temperatura: int) -> bool:
    """Retorna True si hay temperaturas peligrosas."""
    return temperatura >= 32

def alerta_sequia(humedad: Humedad, temperatura: int) -> bool:
    """Retorna True si hay sequía extrema."""
    return humedad == "muy_baja" and temperatura > 30

def activar_riego(humedad: Humedad, pronostico_lluvia: bool, hora: int, temperatura: int) -> bool:
    """Determina si se debe activar el riego."""
    return (
        humedad == "baja" and 
        not pronostico_lluvia and 
        hora_adecuada(hora) and 
        not alerta_calor(temperatura)
    )

def recomendacion(humedad: Humedad, temperatura: int, hora: int, pronostico_lluvia: bool) -> str:
    """Genera una recomendación basada en los sensores."""
    if activar_riego(humedad, pronostico_lluvia, hora, temperatura) and alerta_calor(temperatura):
        return "ALERTA: Temperatura peligrosa. RECOMENDACION: Postergue riego o use sistema por goteo nocturno"
    elif alerta_sequia(humedad, temperatura):
        return "ALERTA CRITICA: Sequia extrema detectada. RECOMENDACION: Riego de emergencia inmediato"
    elif activar_riego(humedad, pronostico_lluvia, hora, temperatura):
        return "RIEGO AUTORIZADO: Condiciones óptimas detectadas"
    else:
        return "SISTEMA EN ESPERA: No se requieren acciones"

def estado_sensores(humedad: Humedad, temperatura: int, hora: int, pronostico_lluvia: bool):
    """Imprime el estado actual de los sensores."""
    print("\n=== ESTADO ACTUAL DE SENSORES ===")
    print(f"Humedad del suelo: {humedad}")
    print(f"Temperatura: {temperatura}°C")
    print(f"Hora actual: {hora}:00")
    print("Pronóstico: " + ("Lluvia esperada" if pronostico_lluvia else "No hay lluvia"))
    print("==================================\n")

# Ejemplo de ejecución
if __name__ == "__main__":
    h = "baja"
    t = 35
    hr = 20
    lluvia = False
    estado_sensores(h, t, hr, lluvia)
    print(recomendacion(h, t, hr, lluvia))
